src/crypto.ts

```diff
@@ -0,0 +1,427 @@
+			provider = <CryptoProvider> _provider;
+		}
+	}
+}
+
+/**
+ * An object for hashing a data stream.
+ */
+export interface Hasher<T extends Data> extends Sink<T> {
+	digest: Promise<ByteBuffer>;  // read only
+}
+
+/**
+ * Call a method that returns a promise, or return a resolved Promise if the method doesn't exist on the object
+ */
+function callOrNoop(object: any, methodName: string, ...args: any[]): Promise<any> {
+	const method: Function = object[methodName];
+	if (!method) {
+		return Promise.resolve()
+	}
+	return method.apply(object, args);
+}
+
+function bindOrUndefined(object: any, methodName: string): any {
+	const method: Function = object[methodName];
+	if (method) {
+		return method.bind(object);
+	}
+}
+
+/**
+ * A wrapper around a Promise<HashFunction> that will defer calls while a provider is asynchronously loaded.
+ */
+class HasherWrapper<T extends Data> implements Hasher<T> {
+	constructor(hashPromise: Promise<HashFunction>, codec: Codec) {
+		Object.defineProperty(this, '_promise', {
+			value: new Promise((resolve, reject) => {
+				hashPromise.then(
+					(hashFunction) => {
+						// When the hash function resolves, create a Hasher and replace this object's methods with
+						// pointers to the corresponding methods on the Hasher.
+						const hasher = hashFunction.create(codec);
+						this.abort = bindOrUndefined(hasher, 'abort');
+						this.close = bindOrUndefined(hasher, 'close');
+						this.start = bindOrUndefined(hasher, 'start');
+						this.write = bindOrUndefined(hasher, 'write');
+						resolve(hasher);
+					},
+					function (error) {
+						reject(error);
+					}
+				);
+			})
+		});
+
+		Object.defineProperty(this, 'digest', {
+			enumerable: true,
+			value: new Promise((resolve, reject) => {
+				this._promise.then(
+					function (hasher) {
+						resolve(hasher.digest);
+					},
+					function (error) {
+						reject(error);
+					}
+				);
+			})
+		});
+	}
+
+	private _promise: Promise<Hasher<T>>;
+
+	digest: Promise<ByteBuffer>;
+
+	// Sink methods; the provider may or may not implement these, so call them with callOrNoop
+
+	abort(reason?: Error): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'abort', reason);
+		});
+	}
+
+	close(): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'close');
+		});
+	}
+
+	start(error: (error: Error) => void): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'start', error);
+		});
+	}
+
+	write(chunk: T): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return callOrNoop(hasher, 'write', chunk);
+		});
+	}
+}
+
+/**
+ * A cryptographic key.
+ */
+export interface Key {
+	algorithm: string,
+	data: Data
+}
+
+/**
+ * An object for signing a data stream.
+ */
+export interface Signer<T extends Data> extends Sink<T> {
+	signature: Promise<ByteBuffer>;  // read only
+}
```

Comment:
might want to move this to the top of the file since it is first used in the `SignFunction` interface definition